### ISSUE TITLE
[feature] file/directory/pattern pinning

### ISSUE THREAD

**Issue by borkd** [moosefs/moosefs#168]
## Have you read through available documentation and open Github issues?

Yes

## Is this a BUG report, FEATURE request, or a QUESTION? Who is the indended audience?
FEATURE request

## What and why?

Storage classes offer tremendous flexibility, but there are some corner cases when admin's work could be made easier by allowing them to administratively pin a file/directory/pattern to a particular set of chunkserver labels, while letting the end user or operator to recirsively change classes for remaining data within the limits set in `mfsexports.cfg`. Basically having an admin set storage class for a directory/file/pattern that overrides whatever user/operator would normally be allowed to accomplish.

Use cases:
- Hostpot mitigation. Directories full of small and frequently changing files where access patterns do create hot spots, slowing down response times for everyone and everything using that chunkserver.
- Wearout mitigation and network load management. Very large, mostly static datasets, where replication created by erroneous class changes would result in unnecessary wear and tear on the rest of the infrastructure. Think namespaces where (create==keep==archive) is intermixed with (create=>keep=>archive) and (create=>(keep==archive)).

**Comment by oxide94** [comment]
Hi @borkd,
Patters support is already implemented in MooseFS 4.x. Patterns can be used to override storage classes, trash retention and/or extra attributes whenever a new file is created or an existing file is renamed.

See attached piece `mfscli` / MFS CGI and manpage for new `mfspatadmin` tool, it may give you an idea on how it is implemented now:

* [`mfspatadmin` manpage](https://github.com/moosefs/moosefs/files/2746975/man_mfspatadmin.txt)

* `mfscli`, piece of RESOURCES section:
```
+---------------------------------------------------------------------------------------+
|                                   Override Patterns                                   |
+---------+------+------+----------+---------------+-----------------+------------------+
| pattern | euid | egid | priority | storage class | trash retention | extra attributes |
+---------+------+------+----------+---------------+-----------------+------------------+
|                                        no data                                        |
+---------------------------------------------------------------------------------------+
```

* MFS CGI, piece of RESOURCES section:
![screen shot 2019-01-10 at 9 14 29 pm](https://user-images.githubusercontent.com/4892057/50995289-fa0f0000-151e-11e9-9f54-442c535142db.png)

Best regards,
Peter / MooseFS Team

**Comment by borkd** [comment]
Yes, I know, and that's why "pattern" made it to issue's title. Folks who evaluate storage solution based on available open-source codebase may have no idea they need to ask for a demo of 4.x, and if they don't need HA may simply walk away because hotspots and/or wear out will manifest themselves on small commodity-based deployments much faster than on true scale-out installations. If public release of 4.x is months or longer away, it may be a good idea to backport that feature.
